### PR TITLE
Add Fragmenter and transport-level fragmentation support

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- Introduces a Fragmenter helper to slice oversized outbound messages into fixed-size transport fragments, each annotated with FragmentHeader metadata (MessageId, FragmentIndex, is_last_fragment).
- Adds FragmentFrame and FragmentBatch to represent transport-level fragments and batches, with ergonomic iteration and access.
- Adds FragmentationError to surface encoding or index-overflow errors during fragmentation.
- Updates public API surface to expose new fragmentation types and wiring for tests and behavioural scenarios.

### Core functionality
- New src/fragment/fragmenter.rs implementing Fragmenter with:
  - new(max_fragment_size) and with_starting_id(max_fragment_size, start_id)
  - max_fragment_size(), next_message_id()
  - fragment_message(&M) -> FragmentBatch, fragment_bytes(payload) -> FragmentBatch, fragment_with_id(message_id, payload) -> FragmentBatch
  - Internal fragment construction that respects max_fragment_size, assigns FragmentIndex, and marks is_last_fragment on the final piece
  - Overflow handling via FragmentationError::IndexOverflow
- New FragmentFrame type for a single outbound fragment (header + payload) with header(), payload(), and into_parts()
- New FragmentBatch type representing a collection of fragments for a single logical message with:
  - message_id(), fragments(), len(), is_empty(), is_fragmented(), into_fragments()
  - IntoIterator implementation for easy consumption
- New FragmentationError enum (in fragment error module):
  - Encode(EncodeError) for serialization failures
  - IndexOverflow { last: FragmentIndex } for fragment index overflow

### Public API surface
- Expose FragmentBatch, FragmentFrame, Fragmenter, and FragmentationError from the fragment module
- Re-export new items in src/lib.rs to form a cohesive fragmentation API surface
- Tests and behavioural tests updated to exercise the new API surface

### Tests
- Expanded unit tests for fragmentation:
  - fragmenter_splits_payload_into_multiple_frames
  - fragmenter_handles_empty_payload
  - fragmenter_fragments_messages_and_increments_ids
  - fragment_batch_into_iterator_yields_all_fragments
- Updated tests/worlds/fragment.rs with Fragmenter configuration, fragmentation requests, and assertions
- Extended tests/features/fragment.feature with scenario: Fragmenter splits oversized payloads into sequential fragments
- Extended tests/steps/fragment_steps.rs with Given/When/Then steps for fragmentation and assertions

### Documentation & changelog
- CHANGELOG.md: note added Fragmenter and transport-level fragmentation capability
- Documentation updates across the docs folder to reflect fragmentation design, including a new Fragmenter section and updated roadmap notes
- User guide (docs/users-guide.md) includes an example demonstrating Fragmenter usage

## Backwards compatibility
- Fragmentation is opt-in via the Fragmenter helper; existing code paths remain unaffected if fragmentation is not used.
- Public API surface has new items but existing APIs remain stable; users may opt into fragmentation by using Fragmenter and FragmentBatch/FragmentFrame types

## Example usage
```rust
use std::num::NonZeroUsize;
use wireframe::fragment::{Fragmenter, FragmentBatch};

let fragmenter = Fragmenter::new(NonZeroUsize::new(512).unwrap());
let batch: FragmentBatch = fragmenter
    .fragment_bytes(&vec![0u8; 1400])
    .expect("fragmentation should succeed");

assert_eq!(batch.len(), 3);
for (i, f) in batch.into_iter().enumerate() {
    println!("frag {}: {} bytes, last: {}", i, f.payload().len(), f.header().is_last_fragment());
}
```

## Test plan
- [x] Run cargo test (unit tests for fragmentation)
- [x] Run cucumber-based behavioural tests for fragmentation scenarios
- [x] Validate edge cases: empty payload, multi-fragment payloads, and index overflow scenarios

🌿 Generated by [Terry](https://www.terragonlabs.com)

---

ℹ️ Tag @terragon-labs to ask questions and address PR feedback

📎 **Task**: https://www.terragonlabs.com/task/b73c1632-521c-4ee5-9a0a-045c2128dbbf

## Summary by Sourcery

Introduce an outbound fragmentation helper and wire it into the public fragment API, along with tests and documentation describing its behavior.

New Features:
- Add a Fragmenter helper that splits outbound messages or raw payloads into fixed-size transport fragments with sequential MessageIds and fragment indices.
- Introduce FragmentFrame and FragmentBatch types to represent individual fragments and batches for a single logical message, supporting ergonomic access and iteration.
- Expose a FragmentationError type to surface encoding failures and fragment index overflows during fragmentation.

Enhancements:
- Extend fragment-world test utilities and Cucumber steps to configure a Fragmenter, request fragmentation, and assert fragment counts, payload sizes, final flags, and message IDs.
- Update documentation to describe the Fragmenter helper, its API, and how it fits into the broader fragmentation and outbound messaging designs, and mark the roadmap item for Fragmenter as completed.

Documentation:
- Document the Fragmenter, FragmentBatch, and FragmentFrame types in the design docs and user guide, including example usage and updated design decisions.
- Adjust markdown linting directives in several design documents to accommodate requirement tables without warnings.

Tests:
- Add unit tests covering Fragmenter behavior for multi-fragment payloads, empty payloads, message ID sequencing, and FragmentBatch iteration.
- Add a new behavioural scenario validating that the Fragmenter splits oversized payloads into sequential fragments with correct sizes, final flags, and message IDs.

Chores:
- Update the changelog to note the new Fragmenter helper and transport-level fragmentation capability.

### CODE REVIEW

File: src/fragment/fragmenter.rs [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+    }
+
+    /// Return the maximum fragment payload size in bytes.
+    #[must_use]
+    pub const fn max_fragment_size(&self) -> NonZeroUsize { self.max_fragment_size }
+
+    /// Generate and return the next [`MessageId`].
+    #[must_use]
+    pub fn next_message_id(&self) -> MessageId {
+        MessageId::new(self.next_message_id.fetch_add(1, Ordering::Relaxed))
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/fragment/fragmenter.rs]: Has this now been resolved?

File: src/fragment/tests.rs [hunk 2]
```diff
@@ -77,3 +80,76 @@ fn series_accepts_final_fragment_at_max_index() {
...
+    assert_eq!(fragment.payload(), &[]);
+    assert!(fragment.header().is_last_fragment());
+    assert_eq!(fragment.header().fragment_index(), FragmentIndex::zero());
+}
+
+#[derive(Debug, Encode, BorrowDecode)]
+struct DummyMessage(Vec<u8>);
+
+#[test]
+fn fragmenter_fragments_messages_and_increments_ids() {
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/fragment/tests.rs]: Has this now been resolved?

File: src/fragment/fragmenter.rs [hunk 3]
```diff
@@ -0,0 +1,201 @@
+//! Outbound helper that splits logical messages into transport fragments.
+//!
+//! [`Fragmenter`] exposes a small API for chunking serialized messages into
+//! fixed-size fragments, tagging each fragment with a [`FragmentHeader`]. The
+//! struct tracks unique [`MessageId`] values internally so callers can request
+//! chunking without worrying about identifier collisions.
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/fragment/fragmenter.rs]: @coderabbitai "serialized" is correct for en-GB-oxendict (-ize / -yse / -our)

File: src/fragment/tests.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (leynos) [reply src/fragment/tests.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API.

File: docs/users-guide.md [hunk 4]
```diff
@@ -148,6 +148,31 @@ re-assembled message is surfaced to handlers. Behavioural tests can reuse the
...
+
+for fragment in batch.fragments() {
+    tracing::info!(
+        msg_id = fragment.header().message_id().get(),
+        index = fragment.header().fragment_index().get(),
+        final = fragment.header().is_last_fragment(),
+        payload_len = fragment.payload().len(),
+    );
+}
+```
```
[the remark this answers is not in the document]
Comment (leynos) [reply docs/users-guide.md]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API.
